### ISSUE TITLE
pdbbind scaffold test set appears to be truncated

### ISSUE THREAD

**Issue by dkoes** [deepchem/deepchem#741]
deepchem/contrib/atomicconv/acnn/refined/get_acnn_refined.sh
```
     test = dc.data.DiskDataset("datasets/scaffold_test")                                                            
     print len(test.ids)
```

This test set only has 708 entries (others have ~740), and they stop exactly at 8pK (there are no high affinity compounds).

It seems unlikely that high affinity compounds (at precisely the 8.00 cutoff) are missing because of scaffold clustering (as opposed to the list getting truncated somewhere).   It's pretty suboptimal for more than a third of the affinity range to be missing from the test set.

**Comment by rbharath** [comment]
This is indeed pretty strange and we should dig into what's happening here. 

CC @lilleswing @joegomes

**Comment by lilleswing** [comment]
Looking through the code for the scaffold splitter we will always under fill the training set and over fill the test set based on granularity of the scaffold clusters.

```python
def create_scaffold_indices(pdbbind_dir, base_dir):

  frac_train = 0.8
  frac_valid = 0.0

  scaffolds = {}
  y = np.array([0 for val in ids])
  w = np.ones_like(y)
  for ind, pdb_code in enumerate(ids):
    print("Processing %s" % str(pdb_code))
    pdb_subdir = os.path.join(pdbbind_dir, pdb_code)
    mol = compute_ligand_mol(pdb_subdir, pdb_code)
    if mol is not None:
      engine = dc.utils.ScaffoldGenerator(include_chirality=False)
      scaffold = engine.get_scaffold(mol)
      if scaffold not in scaffolds:
        scaffolds[scaffold] = [ind]
      else:
        scaffolds[scaffold].append(ind)
    else:
      print(pdb_code)

  # Sort from largest to smallest scaffold sets
  scaffolds = {key: sorted(value) for key, value in scaffolds.items()}
  scaffold_sets = [
      scaffold_set
      for (scaffold, scaffold_set) in sorted(
          scaffolds.items(), key=lambda x: (len(x[1]), x[1][0]), reverse=True)
  ]
  train_cutoff = frac_train * len(y)
  valid_cutoff = (frac_train + frac_valid) * len(y)
  train_inds, valid_inds, test_inds = [], [], []
  for scaffold_set in scaffold_sets:
    if len(train_inds) + len(scaffold_set) > train_cutoff:
      if len(train_inds) + len(valid_inds) + len(scaffold_set) > valid_cutoff:
        test_inds += scaffold_set
      else:
        valid_inds += scaffold_set
    else:
      train_inds += scaffold_set
  return train_inds, valid_inds, test_inds
```
The scaffold split does not look at the Y values at all when doing the split.  This is definitely an argument to improve the scaffold splitter to some sort of stratified scaffold splitter.

I agree that it is suboptimal for a high affinity compounds to be missing from the test set.

**Comment by dkoes** [comment]
Trying to bump this.  I'd like to evaluate the exact same splits used in the paper (https://arxiv.org/abs/1703.10603), but the ones available to download clearly aren't right.  In addition to the truncated test set, there are 17 examples where train and test structures have **identical** ligands:
```
test | train| sim
 --  |  --  | --
1o5c | 1o5g | 1.000000
3qx8 | 1xk5 | 1.000000
3gxy | 1mfa | 1.000000
3f5k | 1gwv | 1.000000
2qpu | 4uac | 1.000000
4dy6 | 2wzm | 1.000000
3hww | 4oct | 1.000000
4l6t | 4gzx | 1.000000
3old | 1ppi | 1.000000
4w9i | 4w9l | 1.000000
4m7j | 3hzk | 1.000000
2x6x | 2w7y | 1.000000
2zym | 1gwv | 1.000000
3obq | 4j47 | 1.000000
3t60 | 3t70 | 1.000000
1h5v | 1jfh | 1.000000
3k02 | 1k1y | 1.000000
```

And there are 73 (>10%) test examples with >0.9 Tanimoto to a training set ligand.

**Comment by evanfeinberg** [comment]
Thanks for your communication, David! If I may, I'd like to suggest that we
all focus on two different splitting techniques anyway: (1) temporal split
(train on earlier samples, test on later samples) and (2) train on refined
minus core, test on core set. These splits are less user-biased than the
scaffold split and split (2) allows comparison to older reported methods
(e.g., http://onlinelibrary.wiley.com/doi/10.1002/minf.201400132/full ). Of
course I agree it makes sense to ultimately resolve scaffold splitting
issues, though I personally think as a field, splitting methods (1) and (2)
should become the new standard. Just my two cents.

**Comment by dkoes** [comment]
What split makes the most sense depends on what sort of generalization error you are trying to measure.  If you are interested in how well you will generalize to new targets, you should split by targets (which is what we do, with a significant difference in sequence identity).  If you are interested in how well you generalize to new chemotypes, a scaffold split makes sense (although this is a bit tricky; e.g. compounds with different scaffolds may still have the same "warheads").

I don't find the time split or the non-core/core split particularly attractive, but to each their own.  They also aren't particularly amenable to cross-validation or bootstrapping.

**Comment by evanfeinberg** [comment]
Totally agreed re: sequence/target splitting. I have personally found this
to be the most challenging (and therefore, perhaps most useful) to measure
generalization. The nice characteristic of target splitting is that it
tends to correlate with splitting chemotypes as well (enzyme inhibitors
look a whole lot different than orthosteric GPCR ligands, for example).

**Comment by rbharath** [comment]
Closing this old discussion. Feel free to re-open if there are new points to consider.
